docs/design/2020-01-24-collations.md

```diff
@@ -0,0 +1,294 @@
+  * Tertiary Weight i.e. case
+  * PAD / NOPAD
+
+All of them are supported by `text/collate` package of Go, so it is possible to map Go collations to some of UCA-based collations in MySQL like `utf8mb4_unicode_ci`/`utf8mb4_0900_ai_ci`, if we ignore the differences between UCA versions: current `text/collate` uses UCA version `6.2.0` and it is not changable. However, the collations in MySQL are with different UCA versions marked in the names, for example, `utf8mb4_0900_ai_ci` uses version `9.0`.
+
+For non-standard UCA implementations in MySQL, i.e. the `utf8mb4_general_ci`. The implementation depends on our choice to the [Compatibility with MySQL](#compatibility-with-mysql) chapter, if a 100% compatibility of `utf8mb4_general_ci` is chosen, we need to implement it by our hands.
+
+### Collations in TiKV
+
+For all collations supported by TiDB, they should be supported by TiKV as well so that the coprocessors can be pushed-down. Several libraries may be helpful for the work:
+
+  * [Rust unicode_collation](https://docs.rs/unicode-collation/0.0.1/unicode_collation)
+  * [Rust servo/rust-icu](https://github.com/servo/rust-icu)
+  * [C++ std::collate](https://en.cppreference.com/w/cpp/locale/collate)
+  * [C/C++ ICU4C](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c)
+
+The libraries above work for standard UCA only, for all other collations, implementations should be done from nothing.
+
+## Migration Plans
+
+The plan depends on the option chosen in [Compatibility](#compatibility) chapter.
+
+  * For existing TiDB clusters with current binary collations, nothing need to be done if the users are happy with them.
+  * For the potential migrations from MySQL to TiDB:
+     - If the collations used in MySQL have been implemented by TiDB, users from MySQL do not need to care about the collations when mirgrating to TiDB except Compatibility Option 2, in which the those collations need to be updated to their corresponding names.
+     - If there are colltions that are not supported by TiDB yet, users may need to change the them to the supported ones and check if no constraint is broken after the change. The check can be done following the approach mentioned in [this article](https://mysqlserverteam.com/mysql-8-0-collations-migrating-from-older-collations).
+
+## Testing Plans
+
+Here lists possible tests that can be done:
+
+  * The unit / integration tests corresponding to the updates in [Changes to make](#changes-to-make)
+  * The unit / integration tests corresponding to the [implementations](#implementation)
+  * The collations cases that ported from mysql-tests
+  * The copr-pushdown tests of TiDB/TiKV
+
+## Possible Future Works
+
+We have an opportunity to do a much better job than MySQL has done.
+
+A collation, logically, should not be a property of a character set. Since all character sets are subsets of Unicode, collations can apply to Unicode codepoints without needing to be tied to individual character sets.
```

Comment:
This might not be true. For example, users may want to store data in GBK encoding while be able to compare by Chinese characters. GBK encoding lead to less storage space compared to Unicode encoding when storing Chinese characters.